### PR TITLE
llvm: Add recipe for llvm12-libunwind

### PR DESCRIPTION
**Pull request by trungnt2910**
Added a recipe for `llvm12_libunwind` and `llvm12_libunwind_devel`.

A few questions may rise, I'm answering just in case:

## What are `libcxx` and `libcxxabi` doing here?
They are required for `libunwind` to build. (Look at the [`CMakeLists.txt`](https://github.com/llvm/llvm-project/blob/release/12.x/libunwind/CMakeLists.txt#L2) file in `libunwind` source for more information. Similarly, `libcxx` requires `libcxxabi`.)

While included in the source directory, the libraries are not built, due to fears of maintaining a C++ library that conflicts with the GNU `libstdc++` linked to `libroot`. The build is disabled by removing the `CMakeLists.txt` in the corresponding folders.

Patchsets for getting these two libraries building on Haiku are still included in this PR, in case of any future interest. The builds for these libraries have been tested on `x86_64` Haiku.

## ~~Why is only `x86_64` supported?~~
~~Like I said in my previous PR, while libunwind supports a wide range of platforms, Haiku-specific signal frame unwinding is currently only available for x86_64.~~

(Irrelevant, basic unwinding works normally on all archs, signal unwinding is only officially supported by libunwind on Linux x86_64 and arm64, @X547's patch is still broken on Haiku).

`x86_gcc2` is sure to be broken (the same reason why it is broken for the rest of LLVM).

## What does the patchset for libunwind contain?
1. Haiku signal frame step handling done by @X547, as in [this Haiku patch](https://review.haiku-os.org/c/haiku/+/5105).
2. Some more trivial `libunwind` configuration and bugfixes. Some of these bugfixes are already available upstream in 14.0.6 (such as the fixes to `Registers.hpp`).

Also note that libunwind only works for Haiku built using the latest buildtools (the one built on 08/07/2022, with `.eh_frame_hdr`) and gcc11 revision 8 (after #7038 is merged).

If either this PR or #7036 is merged, I'll close the other one.

### CODE REVIEW

**Comment by trungnt2910** [comment]
libunwind has been tested on x86_64 and x86.

The recipe itself has only been tested on x86_64 (building LLVM is a time-consuming process).

The recipe should now be ready for review.

File: sys-devel/llvm/patches/libunwind-12.0.1.patchset [hunk 1]
```diff
@@ -0,0 +1,416 @@
...
+- // printf("sp: %p\n", (void*)sp);
++  // printf("sp: %p\n", (void*)sp);
+ #if defined(_LIBUNWIND_TARGET_X86_64)
+   vregs *regs = (vregs*)(sp + 0x70);
+   //printf("&regs: %p\n", regs);
+- 
++
+   _registers.setRegister(UNW_REG_IP, regs->rip);
+-  _registers.setRegister(UNW_REG_SP, regs->rsp);  
++  _registers.setRegister(UNW_REG_SP, regs->rsp);
```
Comment (waddlesplash) [inline sys-devel/llvm/patches/libunwind-12.0.1.patchset:383]: Can you avoid unnecessary whitespace differences please? (There's a way to get git to `add` only non-whitespace changes, I think. Or you can use `git gui` to do per-hunk staging.)

File: sys-devel/llvm/patches/libunwind-12.0.1.patchset [hunk 2]
```diff
@@ -0,0 +1,416 @@
+From 3e3979c2ae83dd20ead2ae41dff99090a982d274 Mon Sep 17 00:00:00 2001
+From: X512 <danger_mail@list.ru>
+Date: Wed, 16 Mar 2022 07:04:18 +0900
+Subject: libroot: implement Itanium ABI unwind functions by using LLVM
```
Comment (waddlesplash) [inline sys-devel/llvm/patches/libunwind-12.0.1.patchset:4]: Well that's not what this change is doing. How much of these changes are really necessary anyway?

File: sys-devel/llvm/patches/llvm-12.0.1.patchset [hunk 3]
```diff
@@ -177,3 +177,28 @@ index f6d882d..e2a3343 100644
...
+
+diff --git a/utils/gn/secondary/libcxx/include/BUILD.gn b/utils/gn/secondary/libcxx/include/BUILD.gn
+index 2ca495b08f..579c134854 100644
+--- a/utils/gn/secondary/libcxx/include/BUILD.gn
++++ b/utils/gn/secondary/libcxx/include/BUILD.gn
+@@ -219,6 +219,7 @@ copy("include") {
+     "strstream",
+     "__support/android/locale_bionic.h",
+     "__support/fuchsia/xlocale.h",
++    "__support/haiku/xlocale.h",
```
[the remark this answers is not in the document]
Comment (waddlesplash) [reply sys-devel/llvm/patches/llvm-12.0.1.patchset:198]: Yeah, that probably makes the most sense.

File: sys-devel/llvm/patches/libunwind-12.0.1.patchset (hunk 2 above)
Comment (trungnt2910) [reply sys-devel/llvm/patches/libunwind-12.0.1.patchset:4]: This is a patch extracted from X512's work and I'm keeping everything intact.

The patch adds support for libunwind signal frame unwinding on Haiku.

Comment (trungnt2910) [reply sys-devel/llvm/patches/libunwind-12.0.1.patchset]: I've added a more meaningful commit message.

File: sys-devel/llvm/patches/libunwind-12.0.1.patchset (hunk 1 above)
Comment (trungnt2910) [reply sys-devel/llvm/patches/libunwind-12.0.1.patchset]: Done.

**Comment by trungnt2910** [comment]
Tested this recipe again, builds and runs well without `libcxx` and `libcxxabi` patches.
